### ISSUE TITLE
PyTorch模型转换后，LogSoftmax 算子计算精度较低（Pytorch->ONNX->MNN）

### ISSUE THREAD

**Issue by 369zz** [alibaba/MNN#1585]
PyTorch 的 LogSoftmax 计算结果和 MNN 的不一致，当输入数据的范围较大时，MNN 会将数值较小的部分统一计算为 -87，丢失了数值间的大小关系。

LogSoftmax 计算过程如下所示，
![image](https://user-images.githubusercontent.com/22151569/126855890-b448521f-216f-4c2a-9d66-a0121fe8a0a5.png)

根据图中高精度实现的公式，将 LogSoftmax 算子拆分为数值运算，这样导出的 MNN 模型计算精度和 PyTorch 就一致了，建议改进一下 MNN 中 LogSoftmax 算子的实现。（另外，顺便把 LogSumExp 这个算子也支持一下呗）

# 平台(如果交叉编译请再附上交叉编译目标平台):  
Ubuntu 18.04 x86

# Github版本:
8a82f5c5a7ae37192784c2dbd6dfc8ca8833565a

# 编译方式:
官方示例

# 编译日志:
无异常

**Comment by yuyq96** [comment]
请问有修改后的代码吗？我也遇到了这个问题，MNNExp和MNNSoftmax的精度损失有点大

**Comment by 369zz** [comment]
> 请问有修改后的代码吗？我也遇到了这个问题，MNNExp和MNNSoftmax的精度损失有点大

为了实现简单，我是在 PyTorch 中手动实现了 log_softmax，这样转 MNN 的时候就不会将其当做一个 LogSoftmax 算子，而是当做几个算子串联，代码如下。（如果要 softmax 的话，外面再套一个 exp 应该就可以了）
```
def log_softmax(x, dim=-1):
    # LogSoftMax(x_i) = x_i - LogSumExp(x)
    x = x - x.max(dim=dim, keepdim=True)[0]
    x = x - torch.log(torch.sum(torch.exp(x), dim=dim, keepdim=True))
    return x
```

**Comment by yuyq96** [comment]
> > 请问有修改后的代码吗？我也遇到了这个问题，MNNExp和MNNSoftmax的精度损失有点大
> 
> 为了实现简单，我是在 PyTorch 中手动实现了 log_softmax，这样转 MNN 的时候就不会将其当做一个 LogSoftmax 算子，而是当做几个算子串联，代码如下。（如果要 softmax 的话，外面再套一个 exp 应该就可以了）
> 
> ```
> def log_softmax(x, dim=-1):
>     # LogSoftMax(x_i) = x_i - LogSumExp(x)
>     x = x - x.max(dim=dim, keepdim=True)[0]
>     x = x - torch.log(torch.sum(torch.exp(x), dim=dim, keepdim=True))
>     return x
> ```

谢谢，我尝试一下这个思路

**Comment by jxt1234** [comment]
已经对应修改了 mnn 的转换 logsoftmax 的代码
